pkg/osvscanner/internal/imagehelpers/imagehelpers.go

```diff
@@ -0,0 +1,137 @@
+package imagehelpers
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/google/osv-scalibr/artifact/image/layerscanning/image"
+	"github.com/google/osv-scalibr/extractor/filesystem/os/osrelease"
+	"github.com/google/osv-scanner/internal/clients/clientinterfaces"
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/reporter"
+)
+
+func BuildImageMetadata(img *image.Image, baseImageMatcher clientinterfaces.BaseImageMatcher) (*models.ImageMetadata, error) {
+	chainLayers, err := img.ChainLayers()
+	if err != nil {
+		// This is very unlikely, as if this would error we would have failed the initial scan
+		return nil, err
+	}
+	m, err := osrelease.GetOSRelease(chainLayers[len(chainLayers)-1].FS())
+	OS := "Unknown"
+	if err == nil {
+		OS = m["PRETTY_NAME"]
+	}
+
+	layerMetadata := []models.LayerMetadata{}
+	for _, cl := range chainLayers {
+		layerMetadata = append(layerMetadata, models.LayerMetadata{
+			DiffID:  cl.Layer().DiffID(),
+			Command: cl.Layer().Command(),
+			IsEmpty: cl.Layer().IsEmpty(),
+		})
+	}
+
+	var baseImages [][]models.BaseImageDetails
+
+	if baseImageMatcher != nil {
+		baseImages, err = baseImageMatcher.MatchBaseImages(context.Background(), layerMetadata)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query for container base images: %w", err)
+		}
+	} else {
+		baseImages = [][]models.BaseImageDetails{
+			// The base image at index 0 is a placeholder representing your image, so always empty
+			// This is the case even if your image is a base image, in that case no layers point to index 0
+			{},
+		}
+	}
+
+	imgMetadata := models.ImageMetadata{
+		OS:            OS,
+		LayerMetadata: layerMetadata,
+		BaseImages:    baseImages,
+	}
+
+	return &imgMetadata, nil
+}
+
+func ExportDockerImage(r reporter.Reporter, dockerImageName string) (string, error) {
+	// Skip saving if the file is already a tar archive.
+	if strings.Contains(dockerImageName, ".tar") {
+		if _, err := os.Stat(dockerImageName); err == nil {
+			return dockerImageName, nil
+		}
+	}
+
+	tempImageFile, err := os.CreateTemp("", "docker-image-*.tar")
+	if err != nil {
+		r.Errorf("Failed to create temporary file: %s\n", err)
+		return "", err
+	}
+
+	err = tempImageFile.Close()
+	if err != nil {
+		return "", err
+	}
+
+	// Check if image exists locally, if not, pull from the cloud.
+	r.Infof("Checking if docker image (%q) exists locally...\n", dockerImageName)
+	cmd := exec.Command("docker", "images", "-q", dockerImageName)
+	output, err := cmd.Output()
+	if err != nil || string(output) == "" {
+		r.Infof("Image not found locally, pulling docker image (%q)...\n", dockerImageName)
+		err = runCommandLogError(r, "docker", "pull", "-q", dockerImageName)
+		if err != nil {
+			return "", fmt.Errorf("failed to pull container image: %w", err)
```

Comment:
should we be deleting the temp dir in these error paths? the caller has no way to do it either since we're returning empty string in these cases.

Reply:
Ah good idea, done.